### PR TITLE
fix cardinality, #36119, #36118, #36116

### LINKED ISSUE #36116: Surprising behaviour of PlanePartitions with given symmetry.
### Steps To Reproduce

`PlanePartitions` silently ignores the `symmetry` keyword when the first argument is an integer:
```
sage: [PlanePartitions(n, symmetry="TSSCPP").cardinality() for n in range(15)]
[1, 1, 3, 6, 13, 24, 48, 86, 160, 282, 500, 859, 1479, 2485, 4167]
sage: oeis(_)
0: A000219: Number of planar partitions (or plane partitions) of n.
```

For `TSSCPP` this is particularly puzzling, since in this case, the box must be a cube, so we would have to do
```
sage: [PlanePartitions([n,n,n], symmetry="TSSCPP").cardinality() for n in range(0,15,2)]
[1, 1, 2, 7, 42, 429, 7436, 218348]
```
instead.  As a side note, it would be more useful if `PlanePartitions(3, symmetry="TSSCPP").cardinality()` would return zero instead of raising an error.

### Expected Behavior

I would expect
```
sage: [PlanePartitions(n, symmetry="TSSCPP").cardinality() for n in range(15)]
[1, 0, 1, 0, 2, 0, 7, 0, 42, 0, 429, 0, 7436, 0, 218348]
```

### Actual Behavior

```
sage: [PlanePartitions(n, symmetry="TSSCPP").cardinality() for n in range(15)]
[1, 1, 3, 6, 13, 24, 48, 86, 160, 282, 500, 859, 1479, 2485, 4167]
```

### Additional Information

_No response_

### Environment

```markdown
irrelevant
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

### LINKED ISSUE #36118: PlanePartitions are not hashable
### Steps To Reproduce

```
sage: p = PlanePartitions(4).an_element(); p
Plane partition [[4]]
sage: hash(p)
Traceback (most recent call last):
...
TypeError: unhashable type: 'list'
```

### Expected Behavior

```
sage: p = PlanePartitions(4).an_element()
sage: hash(p)
```
should return a hash.

### Actual Behavior

A `TypeError` is raised.

### Additional Information

Plane partitions inherit from `ClonableArray`, but the elements of the array are lists, eg.:
```
    def __init__(self, parent, pp, check=True):
...
        if isinstance(pp, PlanePartition):
            ClonableArray.__init__(self, parent, pp, check=False)
        else:
            pp = [list(_) for _ in pp]
            if pp:
                for i in reversed(range(len(pp))):
                    while pp[i] and not pp[i][-1]:
                        del pp[i][-1]
                    if not pp[i]:
                        pp.pop(i)
            ClonableArray.__init__(self, parent, pp, check=check)
...
```

### Environment

```markdown
irrelevant.
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

### LINKED ISSUE #36119: not all TSSCPP are produced
### Steps To Reproduce

```
sage: B = PlanePartitions([6,6,6], symmetry="TSSCPP")
sage: list(B)
[Plane partition [[6, 6, 6, 3, 3, 3], [6, 6, 6, 3, 3, 3], [6, 6, 6, 3, 3, 3], [3, 3, 3], [3, 3, 3], [3, 3, 3]],
 Plane partition [[6, 6, 6, 5, 5, 3], [6, 5, 5, 3, 3, 1], [6, 5, 5, 3, 3, 1], [5, 3, 3, 1, 1], [5, 3, 3, 1, 1], [3, 1, 1]],
 Plane partition [[6, 6, 6, 3, 3, 3], [6, 6, 6, 3, 3, 3], [6, 6, 6, 3, 3, 3], [3, 3, 3], [3, 3, 3], [3, 3, 3]],
 Plane partition [[6, 6, 6, 4, 3, 3], [6, 6, 6, 4, 3, 3], [6, 6, 4, 3, 2, 2], [4, 4, 3, 2], [3, 3, 2], [3, 3, 2]],
 Plane partition [[6, 6, 6, 5, 5, 3], [6, 5, 5, 4, 3, 1], [6, 5, 4, 3, 2, 1], [5, 4, 3, 2, 1], [5, 3, 2, 1, 1], [3, 1, 1]],
 Plane partition [[6, 6, 6, 4, 3, 3], [6, 6, 6, 4, 3, 3], [6, 6, 4, 3, 2, 2], [4, 4, 3, 2], [3, 3, 2], [3, 3, 2]],
 Plane partition [[6, 6, 6, 4, 3, 3], [6, 6, 6, 3, 3, 3], [6, 6, 5, 3, 3, 2], [4, 3, 3, 1], [3, 3, 3], [3, 3, 2]]]
sage: B[0] == B[2]
True
```

### Expected Behavior

There should be seven distinct plane partitions in a box with side length 6.

### Actual Behavior

Some plane partitions are missing, others are duplicate.

### Additional Information

_No response_

### Environment

```markdown
irrelevant.
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

### PR DESCRIPTION
**Pull request by mantepse**
* Fix a thinko in the iterator for TSSCPP
* make plane partitions hashable
* avoid rational numbers in the cardinality methods, when necessary
* raise an error when an integer and a symmetry is provided as argument to `PlanePartitions`

Fixes #36119
Fixes #36118 
Fixes #36116 

- [X] The title is concise, informative, and self-explanatory.
- [X] The description explains in detail what this PR is about.
- [X] I have linked a relevant issue or discussion.
- [X] I have created tests covering the changes.

### CODE REVIEW

File: src/sage/combinat/plane_partition.py [hunk 1]
```diff
@@ -1708,6 +1722,11 @@ def __iter__(self) -> Iterator:
 
             sage: list(PlanePartitions(2))
             [Plane partition [[2]], Plane partition [[1, 1]], Plane partition [[1], [1]]]
+
+        TESTS::
+
+            sage: all(len(set(PP)) == PP.cardinality() for n in range(1, 10) if (PP := PlanePartitions(n)))
```
Comment (fchapoton) [inline src/sage/combinat/plane_partition.py]: please use a smaller upper bound here

File: src/sage/combinat/plane_partition.py [hunk 2]
```diff
@@ -1922,6 +1941,11 @@ def __iter__(self) -> Iterator:
             Plane partition [[1, 1], [1, 1]],
             Plane partition [[1, 1], [1]],
             Plane partition [[1]]]
+
+        TESTS::
+
+            sage: all(len(set(PP)) == PP.cardinality() for n in range(1, 5) if (PP := PlanePartitions([n]*3, symmetry='SPP')))
+            True
```
Comment (fchapoton) [inline src/sage/combinat/plane_partition.py:2007]: lower upper bound here too

**Comment by fchapoton** [comment]
The cardinality doctests are not really useful. To avoid spending time on that tests, please use a very small bound or test case.

**Comment by mantepse** [comment]
> well, that's fast but we must keep in mind that sage is currently taking more than 3 hours to run the doctests, and this is a pain

Is it (significantly) slower on your computer?

I now see that the results are cached, but also the first run seems to be super fast.

I feel a bit uneasy to test only 5 numbers (1,1,2,7,42) for TSSCPP, and the test for all plane partitions is by far the fastest.

I just noticed that I should also test `n=0`.

However, the really slow test is
```
sage: list(PlanePartitions([4,4,2], symmetry='SSCPP'))
```
which takes a second.  I think I have a good fix for that.  Just a minute.

**Comment by mantepse** [comment]
I think this is quite an improvement now.

File: src/sage/combinat/plane_partition.py [hunk 3]
```diff
@@ -1632,7 +1699,11 @@ def cardinality(self) -> Integer:
         A = self._box[0]
         B = self._box[1]
         C = self._box[2]
-        return Integer(prod(Integer(i + j + k - 1) / Integer(i + j + k - 2)
+        return Integer(prod(i + j + k - 1
+                            for i in range(1, A + 1)
+                            for j in range(1, B + 1)
+                            for k in range(1, C + 1)) //
+                       prod(i + j + k - 2
```
Comment (fchapoton) [inline src/sage/combinat/plane_partition.py:1706]: potentially, this is blowing up faster by not simplifying on the run

**Comment by fchapoton** [comment]
about the changes in the product formulas. Mostly good, but one could use `ZZ.prod` to get sage integers.

Also I wonder if separating products for numerator and denominator could lead to huge intermediate numbers

**Comment by mantepse** [comment]
> about the changes in the product formulas. Mostly good, but one could use `ZZ.prod` to get sage integers.
> 
> Also I wonder if separating products for numerator and denominator could lead to huge intermediate numbers

I agree in principle, but isn't this premature optimization?

For example,
```
sage: for a, b in cartesian_product([range(100)]*2):
....:     P = PlanePartitions([a, a, b], symmetry='SPP')
....:     (a, b, P.cardinality())
```
works without any problems.

**Review by fchapoton — APPROVED** [review]
looks good now. Should I set to positive review ?

**Pull request merged**